File: src/ConversationListAdapter.scala
package org.aprsdroid.app

import _root_.android.app.Activity
import _root_.android.content._
import _root_.android.database.Cursor
import _root_.android.os.{AsyncTask, Bundle, Handler}
import _root_.android.text.format.DateUtils
import _root_.android.util.Log
import _root_.android.view.View
import _root_.android.widget.{SimpleCursorAdapter, TextView}
import _root_.androidx.core.content.ContextCompat

object ConversationListAdapter {
	import StorageDatabase.Message._
	val LIST_FROM = Array(CALL, TEXT)
	val LIST_TO = Array(R.id.call, R.id.message)
}

class ConversationListAdapter(context : Context, prefs : PrefsWrapper)
		extends SimpleCursorAdapter(context, R.layout.conversationview, null,
			ConversationListAdapter.LIST_FROM, ConversationListAdapter.LIST_TO) {

	lazy val storage = StorageDatabase.open(context)

	reload()

	lazy val locReceiver = new LocationReceiver2(load_cursor,
		replace_cursor, cancel_cursor)

	ContextCompat.registerReceiver(context, locReceiver, new IntentFilter(AprsService.MESSAGE), Context.RECEIVER_EXPORTED)

	override def bindView(view : View, context : Context, cursor : Cursor) {
		import StorageDatabase.Message._
		val ts = cursor.getLong(COLUMN_TS)
		view.findViewById(R.id.call).asInstanceOf[TextView].setTextColor(0xff00677d)
		view.findViewById(R.id.message).asInstanceOf[TextView].setTextColor(0xff40484c)
		val age = DateUtils.getRelativeTimeSpanString(context, ts)
		view.findViewById(R.id.ts).asInstanceOf[TextView].setText(age)
		super.bindView(view, context, cursor)
	}

	def load_cursor(i : Intent) = {
		val c = storage.getConversations()
		c.getCount()
		c
	}

	def replace_cursor(c : Cursor) {
		changeCursor(c)
		context.asInstanceOf[LoadingIndicator].onStopLoading()
	}
	def cancel_cursor(c : Cursor) {
		c.close()
	}

	def reload() {
		locReceiver.startTask(null)
	}

	def onDestroy() {
		context.unregisterReceiver(locReceiver)
		changeCursor(null)
	}
}

File: src/ConversationListAdapter.kt
package org.aprsdroid.app

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.database.Cursor
import android.text.format.DateUtils
import android.view.View
import android.widget.SimpleCursorAdapter
import android.widget.TextView
import androidx.core.content.ContextCompat

class ConversationListAdapter(
    private val context: Context,
    prefs: PrefsWrapper
) : SimpleCursorAdapter(
    context, R.layout.conversationview, null,
    LIST_FROM, LIST_TO
) {

    companion object {
        val LIST_FROM = arrayOf(StorageDatabase.Companion.Message.CALL, StorageDatabase.Companion.Message.TEXT)
        val LIST_TO = intArrayOf(R.id.call, R.id.message)
    }

    val storage: StorageDatabase by lazy { StorageDatabase.open(context) }

    val locReceiver = LocationReceiver2(
        { loadCursor() },
        { replaceCursor(it) },
        { cancelCursor(it) }
    )

    @SuppressLint("WrongConstant")
    private fun registerReceiver() {
        ContextCompat.registerReceiver(context, locReceiver, IntentFilter(AprsService.MESSAGE), ContextCompat.RECEIVER_EXPORTED)
    }

    init {
        reload()
        registerReceiver()
    }

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        val ts = cursor.getLong(StorageDatabase.Companion.Message.COLUMN_TS)
        view.findViewById<TextView>(R.id.call).setTextColor(0xff00677d.toInt())
        view.findViewById<TextView>(R.id.message).setTextColor(0xff40484c.toInt())
        val age = DateUtils.getRelativeTimeSpanString(context, ts)
        view.findViewById<TextView>(R.id.ts).text = age
        super.bindView(view, context, cursor)
    }

    fun loadCursor(): Cursor {
        val c = storage.getConversations()
        c.count
        return c
    }

    fun replaceCursor(c: Cursor) {
        changeCursor(c)
        (context as? LoadingIndicator)?.onStopLoading()
    }

    fun cancelCursor(c: Cursor) {
        c.close()
    }

    fun reload() {
        locReceiver.startTask(Intent())
    }

    fun onDestroy() {
        try { context.unregisterReceiver(locReceiver) } catch (_: Exception) {}
        changeCursor(null)
    }
}
